Dynamic SSL buffers

AdityaHPatwardhan opened the pull request:
Dynamic SSL buffers

This commit adds changes from the dynamic_ssl_buffers PR(https://github.com/ARMmbed/mbedtls/pull/1208)

    Co-authored-by: Aditya Patwardhan <aditya.patwardhan@espressif.com>
    Revert the breaking changes introduced by the dynamic SSL buffers PR.
    Includes some required changes in the above PR (for rebasing).

## Description
Original work done in [PR](https://github.com/ARMmbed/mbedtls/pull/1208) by Andy Green <andy@warmcat.com> , I have rebased and addressed review comments to keep change backward compatible. The PR helps to reduce the amount of heap simultaneously used by mbedTLS. With local testing on esp32 I have noticed that the difference between minimum_heap_used with and w/o dynamic ssl buffers was considerable.

* the following results were obtained by running the [mqtt/ssl_mutual_auth](https://github.com/espressif/esp-idf/tree/master/examples/protocols/mqtt/ssl_mutual_auth)  example on esp32.

Status| Minimum free heap size
--|--
Without Dynamic SLL buffers(current)|  180472 bytes
With Dynamic SLL buffers|  196544 bytes


Larger value of  minimum free heap size denotes that the Mbedtls used less memory simultaneously. The memory spared by Mbedtls can be used by other tasks.

## Status
**READY**

## Requires Backporting
NO  


## Migrations
 NO

## Additional comments
Any additional information that could be of interest

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

mpg commented:
Hi @AdityaHPatwardhan and thanks for your contribution!

Before taking a closer look at the details of the patch, there's a few points I'd like to discuss with you:

1. Your PR targets `mbedtls-2.17`, which is not a maintained branch and cannot accept changes. New features go to the `development` branch. The other maintained branches are 2.16.x and 2.7.x, which are LTS branches and only accept fixes bor bugs an security issues. (By the way, if you're using 2.17, you should switch to one of the maintained branches as soon as possible, since 2.17 has publicly-known vulnerabilities.)

2. In 2.22.0 we added a new feature, controlled by the config option `MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH`, which is different from what I think your PR is doing, but related in that it shares a common goal: saving RAM used by SSL connections. Did you have a look at that feature? Could you elaborate on how your PR differs from it (what are the pros and cons of each approach)?

3. Also, it's worth noting that we are planning to rewrite most of the message processing stack, and while that rewrite is pending, any change in that area runs the risk of duplicating work, so we'd like to avoid working on changes if know they're going to become obsolete when the rewrite comes in. @hanno-arm what do you think of this specific case - does it conflict with or duplicate planned work?

AdityaHPatwardhan commented:
@mpg Thank you for the reply and clarification.

> 1. Your PR targets `mbedtls-2.17`, which is not a maintained branch and cannot accept changes. New features go to the `development` branch. 

Yes I understand the target branch here is `mbedtls-2.17` and it does not accept merge requests. It was never my intention to get it merged here ( My bad that I did not mention this in the description). I just rebased this (really old) PR (https://github.com/ARMmbed/mbedtls/pull/1208)  to a release branch which was easier than rebasing it over the `development` branch. I was hoping to have some discussion regarding the PR i.e. whether this is still useful in Mbedtls and would be allowed in `development` branch after a few changes.

> 2. In 2.22.0 we added a new feature, controlled by the config option `MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH`, which 

No, I have not looked at the feature. Thank you for notifying, I will take a look at the feature and list the points in which this PR is different.

hanno-arm commented:
@AdityaHPatwardhan Thank you for your work and for taking a look at the new feature `MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH`. As @mpg mentioned, we plan to replace the entire messaging layer of Mbed TLS by a rewrite called MPS which is currently developed alongside the TLS 1.3 prototype for Mbed TLS in https://github.com/hannestschofenig/mbedtls/tree/tls13-prototype. The goal is to have a first version of this upstreamed into Mbed TLS by mid-21. MPS will also enhance flexibility of the buffer handling, and in particular allow for dynamic buffer sizes. I'd therefore rather hold back any work in the area of the messaging layer if possible, as there's risk it overlaps with the MPS / TLS 1.3 efforts.

mpg commented:
@AdityaHPatwardhan Thanks for clarifying why 2.17! It is indeed a good idea to allow some time for discussion before you spend too much effort on the PR. Especially since it's a rather special time where we're less likely to accept changes in this area (as Hanno just confirmed). However, if you can analyse the differences between this PR and the feature that was introduced in 2.22, I think that would be valuable.

AdityaHPatwardhan commented:
Thank you @hanno-arm and @mpg for your inputs.

AdityaHPatwardhan commented:
As suggested by the reviewers, this is similar to a feature added in the Mbedtls-2.22.0. The feature `MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH ` serves the purpose of dynamic buffers which is the main goal of this PR. Closing this.
